src/Covenant/Internal/Data.hs

```diff
@@ -0,0 +1,151 @@
+_mapValT :: forall (a :: Type). (ValT a -> ValT a) -> ValT a -> ValT a
+_mapValT f = \case
+  -- for terminal nodes we just apply the function
+  absr@(Abstraction {}) -> f absr
+  bif@BuiltinFlat {} -> f bif
+  -- For CompT and Datatype we apply the function to the components and then to the top level
+  ThunkT (CompT cnt (CompTBody compTargs)) -> f (ThunkT $ CompT cnt (CompTBody (f <$> compTargs)))
+  Datatype tn args -> f $ Datatype tn (_mapValT f <$> args)
+
+-- This tells us whether the ValT *is* a recursive child of the parent type
+isRecursiveChildOf :: TyName -> ValT AbstractTy -> Reader ScopeBoundary Bool
+isRecursiveChildOf tn = \case
+  Datatype tn' args
+    | tn' == tn -> checkArgsIsRec 0 (V.toList args)
+    | otherwise -> pure False
+  _ -> pure False
+  where
+    -- Checks that the arguments to a datatype (which we expect will be checked to have the same tycon as the parent type)
+    -- have the correct DeBruijn index & the same order as in the initial binding context of the declaration.
+    -- Has to be monadic because need access to the current scope level.
+    checkArgsIsRec :: Int -> [ValT AbstractTy] -> Reader ScopeBoundary Bool
+    checkArgsIsRec _ [] = pure True
+    checkArgsIsRec n (x : xs) = case x of
+      Abstraction (BoundAt db varIx) -> do
+        ScopeBoundary here <- ask
+        let dbInt = asInt db
+        -- Explanation: A component ValT is only a recursive instance of the parent type if
+        --              the DeBruijn index of its type variables points to Z (and the other conditions obtain)
+        if dbInt - here == 0 && review intIndex varIx == n
+          then checkArgsIsRec (n + 1) xs
+          else pure False
+      _ -> pure False
+
+allComponentTypes :: DataDeclaration AbstractTy -> [ValT AbstractTy]
+allComponentTypes = toListOf (datatypeConstructors % folded % constructorArgs % folded)
+
+-- This tells us whether a ValT *contains* a direct recursive type. I.e it tells us whether we need to construct a base functor
+hasRecursive :: TyName -> ValT AbstractTy -> Reader ScopeBoundary Bool
+hasRecursive tn = \case
+  Abstraction {} -> pure False
+  BuiltinFlat {} -> pure False
+  -- NOTE: This assumes that we've forbidden higher rank arguments to constructors (i.e. we can ignore the scope here)
+  ThunkT (CompT _ (CompTBody (NEV.toList -> compTArgs))) -> local (+ 1) $ do
+    or <$> traverse (hasRecursive tn) compTArgs
+  dt@(Datatype _ args) -> do
+    thisTypeIsRecursive <- isRecursiveChildOf tn dt
+    aComponentIsRecursive <- or <$> traverse (hasRecursive tn) args
+    pure $ thisTypeIsRecursive || aComponentIsRecursive
+
+-- Yeah yeah I could use ReaderT but this is less awkward imo, we only care about the "maybe" at the top level when detecting recursion
+mkBaseFunctor :: DataDeclaration AbstractTy -> Reader ScopeBoundary (Maybe (DataDeclaration AbstractTy))
+mkBaseFunctor (DataDeclaration tn numVars ctors) = do
+  anyRecComponents <- or <$> traverse (hasRecursive tn) allCtorArgs
```

Comment:
This might be better as a fold, but I won't insist.

Reply:
Wait the whole thing or just the `or <$> traverse blah` part?

Reply:
The whole thing. `or <$> traverse f` is basically `foldM` with a `Bool` accumulator.

Reply:
In the interest of budget I'm just going to leave it, I don't think spending the time to rewrite it is really worth the payoff (and I think how I did it is reasonably clear even if someone could justifiably claim it's not the most idiomatic)